broadway_api/utils/bootstrap.py

```diff
@@ -49,18 +90,110 @@ def initialize_course_tokens(settings, course_config_path):
         course_dao.insert_or_update(course)
 
 
-def initialize_database(settings):
-    uri = "mongodb://{}:{}".format(config.DB_HOST, config.DB_PORT)
+def initialize_database(settings: Dict[str, Any], flags: Dict[str, Any]):
+    logger.info("initializing database")
 
     try:
-        db_client = MongoClient(uri, serverSelectionTimeoutMS=5000)
+        db_client = MongoClient(
+            flags["mongodb_dsn"],
+            serverSelectionTimeoutMS=flags["mongodb_timeout"] * 1000,
+        )
         db_client.server_info()
+
+        settings["DB"] = db_client
+
+        dao = WorkerNodeDao(settings)
+        logger.info("resetting ws worker nodes")
+        dao.reset_worker_nodes()
+
     except ConnectionFailure as e:
         logger.critical("failed to connect to mongo server: {}".format(repr(e)))
         sys.exit(1)
 
-    settings["DB"] = MongoClient(uri)
 
-    dao = WorkerNodeDao(settings)
-    logger.info("resetting ws worker nodes")
-    dao.reset_worker_nodes()
+def initialize_signal_handler(settings: Dict[str, Any], flags: Dict[str, Any]):
+    logger.info("initializing signal handler")
+
+    def shutdown():
```

Comment:
Now that I think of it, this would not completely shutdown the API. At the time this was written, the API was completely single-threaded. #63 added new threads which talk to the ws workers. Those threads need to be terminated. I think we should open an issue for this and deal with this separately. Please resolve this comment when an issue is opened.

Reply:
hmm I don't quite understand why the ws handlers would add a new thread.

Reply:
Totally my bad, I thought I saw something about creating a new thread when a ws worker registers. Nvm.